sonic_y_cable/y_cable_base.py

```diff
@@ -0,0 +1,1159 @@
+        API will now return Tor B. It also implies that if the link is actively sending traffic on this port,
+        Y cable. MUX will start forwarding packets from TOR B to NIC, and drop packets from TOR A to NIC
+        regardless of previous forwarding state.
+        The port on which this API is called for can be referred using self.port.
+
+        Args:
+
+        Returns:
+            a Boolean, True if the toggle succeeded and False if it did not succeed.
+        """
+        raise NotImplementedError
+
+    def get_read_side(self):
+        """
+        This API checks which side of the Y cable the reads are actually getting performed
+        from, either TOR A or TOR B or NIC and returns the value.
+        The port on which this API is called for can be referred using self.port.
+
+        Args:
+
+        Returns:
+            One of the following predefined constants:
+                TARGET_TOR_A, if reading the Y cable from TOR A side.
+                TARGET_TOR_B, if reading the Y cable from TOR B side.
+                TARGET_NIC, if reading the Y cable from NIC side.
+                TARGET_UNKNOWN, if reading the Y cable API fails.
+        """
+
+        raise NotImplementedError
+
+    def get_mux_direction(self):
+        """
+        This API checks which side of the Y cable mux is currently point to
+        and returns either TOR A or TOR B. Note that this API should return mux-direction
+        regardless of whether the link is active and sending traffic or not.
+        The port on which this API is called for can be referred using self.port.
+
+        Args:
+
+        Returns:
+            One of the following predefined constants:
+                TARGET_TOR_A, if mux is pointing to TOR A side.
+                TARGET_TOR_B, if mux is pointing to TOR B side.
+                TARGET_UNKNOWN, if mux direction API fails.
+        """
+
+        raise NotImplementedError
+
+    def get_active_linked_tor_side(self):
+        """
+        This API checks which side of the Y cable is actively linked and sending traffic
+        and returns either TOR A or TOR B.
+        The port on which this API is called for can be referred using self.port.
+        This is different from get_mux_direction in a sense it also implies the link on the side
+        where mux is pointing to must be active and sending traffic, whereas get_mux_direction
+        just tells where the mux is pointing to.
+
+        Args:
+
+        Returns:
+            One of the following predefined constants:
+                TARGET_TOR_A, if TOR A is actively linked and sending traffic.
+                TARGET_TOR_B, if TOR B is actively linked and sending traffic.
+                TARGET_UNKNOWN, if checking which side is linked and sending traffic API fails.
+        """
+
+        raise NotImplementedError
+
+    def is_link_active(self, target):
```

Comment:
currently by default if an API fails we return a generic error EEPROM_FAILED. This works well since if the RC is this code, we put an N/A inside the value, otherwise True or False indicates the actual return value.

Reply:
Following on that all API's if they fail can just return a negative error code, which if not defined in base class.
can be subsitituted by EEPROM_ERROR=-1